File: ml/forensics.py
"""Image forensics & evidence trust scoring — pure NumPy/PIL, no external APIs.

Three independent signals per evidence photo, plus the composite trust score:

  1. EXIF integrity     — camera make/model, original timestamp, GPS presence,
                          editing-software tags (Photoshop/GIMP/Snapseed/…),
                          and the "stripped metadata" tell (screenshots,
                          re-saved forwards, messenger uploads).
  2. Error-Level Analysis (ELA) — re-encode as JPEG q90 and diff; tampered
                          regions re-compress at a different error level than
                          the rest of the image. We report mean error, error
                          spread and the number of localized hotspots.
  3. Perceptual hash    — 64-bit DCT pHash so the same photo submitted again
                          (different filenames, sizes) can be recognized.

compose_trust_score() fuses the signals with context (geolocation, crowd
corroboration, source reputation, text-image consistency from the CNN) into a
0–1 trust score with a fully inspectable breakdown — the CivicLens
differentiator: not just classification, evidence *provenance*.

Used by ml/serve.py::/forensics and ::/trust.
"""
from __future__ import annotations

import io
import math
import re
from datetime import datetime, timezone
from typing import Any

import numpy as np
from PIL import Image
from scipy.fft import dctn

# ---------------------------------------------------------------------------
# Editing-software fingerprints (case-insensitive substring match on EXIF
# Software / Make / Model / UserComment tags).
# ---------------------------------------------------------------------------

EDITOR_SOFTWARE = [
    "photoshop", "adobe", "lightroom", "gimp", "snapseed", "canva",
    "picsart", "facetune", "pixelmator", "affinity", "illustrator",
    "after effects", "figma", "paint.net", "paint 3d", "mspaint",
    "krita", "darktable", "polarr", "vsco", "meitu", "b612",
]

# Sources where a citizen identity is known/verified vs anonymous.
SOURCE_REPUTATION: dict[str, float] = {
    "whatsapp": 0.75,
    "email": 0.7,
    "manual": 0.7,
    "x": 0.55,
    "tweet": 0.55,
    "news": 0.8,
    "social": 0.5,
    "bulk": 0.4,
}

PHASH_SIZE = 32
PHASH_LOW = 8


# ---------------------------------------------------------------------------
# Loading & hashing
# ---------------------------------------------------------------------------

def load_image(data: bytes) -> Image.Image:
    """Decode image bytes. Raises ValueError on undecodable input."""
    try:
        img = Image.open(io.BytesIO(data))
        img.load()
        return img
    except Exception as e:  # noqa: BLE001
        raise ValueError(f"undecodable image: {e}") from e


def perceptual_hash(img: Image.Image) -> str:
    """64-bit DCT pHash (hex). Robust to resize, recompression, small crops."""
    g = img.convert("L").resize((PHASH_SIZE, PHASH_SIZE), Image.LANCZOS)
    a = np.asarray(g, dtype=np.float64)
    d = dctn(a, type=2, norm="ortho")[:PHASH_LOW, :PHASH_LOW]
    med = float(np.median(d[1:, 1:]))  # ignore DC term for the median
    bits = (d > med).flatten()
    # pack 64 bits -> 16 hex chars
    value = 0
    for b in bits:
        value = (value << 1) | int(b)
    return f"{value:016x}"


def hamming_distance(h1: str, h2: str) -> int:
    """Bit distance between two hex pHashes (0 = identical evidence)."""
    return bin(int(h1, 16) ^ int(h2, 16)).count("1")


def phash_similarity(h1: str, h2: str) -> float:
    """1.0 = same image, 0.0 = maximally different (64-bit space)."""
    return 1.0 - hamming_distance(h1, h2) / 64.0


# ---------------------------------------------------------------------------
# EXIF integrity
# ---------------------------------------------------------------------------

def _exif_dict(img: Image.Image) -> dict[int, Any]:
    try:
        exif = img.getexif()
        raw = {k: v for k, v in exif.items()}
        # IFD sub-dirs: Exif (34665) holds DateTimeOriginal etc.
        try:
            sub = exif.get_ifd(34665)
            raw.update({k: v for k, v in sub.items()})
            gps = exif.get_ifd(34853)
            if gps:
                raw[34853] = dict(gps)
        except Exception:  # noqa: BLE001 — malformed IFDs must not crash us
            pass
        return raw
    except Exception:  # noqa: BLE001
        return {}


def assess_exif(img: Image.Image, submitted_at: str | None = None) -> dict[str, Any]:
    """EXIF-based authenticity signals with human-readable flags."""
    flags: list[str] = []
    raw = _exif_dict(img)
    make = str(raw.get(271, "") or "").strip()
    model = str(raw.get(272, "") or "").strip()
    software = str(raw.get(305, "") or "").strip()
    date_tag = str(raw.get(36867) or raw.get(306) or "").strip()
    has_gps = bool(raw.get(34853))

    edited_by = None
    blob = f"{software} {make} {model}".lower()
    for editor in EDITOR_SOFTWARE:
        if editor in blob:
            edited_by = editor
            break
    if edited_by:
        flags.append(f"edited_software:{edited_by}")

    has_camera = bool(make or model)
    if not raw:
        flags.append("metadata_stripped")
    elif not has_camera:
        flags.append("no_camera_identity")

    age_days: float | None = None
    future_timestamp = False
    if date_tag:
        for fmt in ("%Y:%m:%d %H:%M:%S", "%Y-%m-%d %H:%M:%S"):
            try:
                taken = datetime.strptime(date_tag[:19], fmt).replace(tzinfo=timezone.utc)
                now = datetime.now(timezone.utc)
                age_days = (now - taken).total_seconds() / 86400.0
                if age_days < -1:
                    future_timestamp = True
                    flags.append("future_timestamp")
                elif submitted_at:
                    sub = datetime.fromisoformat(submitted_at.replace("Z", "+00:00"))
                    skew = (sub - taken).total_seconds() / 86400.0
                    if skew > 60:
                        flags.append("stale_photo:>60d_before_submission")
                    elif skew < -1:
                        flags.append("photo_taken_after_submission")
                break
            except ValueError:
                continue
    elif raw:
        flags.append("no_capture_timestamp")

    if not has_gps:
        flags.append("no_gps_exif")

    # Positive integrity signal: full camera + timestamp + GPS is the
    # profile of an on-device original photograph.
    integrity = 0.0
    if has_camera:
        integrity += 0.3
    if date_tag and age_days is not None and not future_timestamp:
        integrity += 0.3
    if has_gps:
        integrity += 0.2
    if not edited_by:
        integrity += 0.2

    return {
        "make": make or None,
        "model": model or None,
        "software": software or None,
        "captured_at": date_tag or None,
        "captured_age_days": None if age_days is None else round(age_days, 1),
        "has_gps": has_gps,
        "edited_by": edited_by,
        "integrity": round(min(1.0, integrity), 3),
        "flags": flags,
    }


# ---------------------------------------------------------------------------
# Error-Level Analysis
# ---------------------------------------------------------------------------

def error_level_analysis(img: Image.Image) -> dict[str, Any]:
    """Re-encode as JPEG q90 and diff — tampered regions light up.

    Returns mean error, error spread (std) and hotspot concentration.
    A uniformly-compressed original yields low, evenly spread error;
    spliced-in patches produce localized high-error clusters.
    """
    rgb = img.convert("RGB")
    buf = io.BytesIO()
    rgb.save(buf, format="JPEG", quality=90)
    buf.seek(0)
    resaved = Image.open(buf)
    resaved.load()

    a = np.asarray(rgb, dtype=np.float64)
    b = np.asarray(resaved, dtype=np.float64)
    diff = np.abs(a - b)

    mean_err = float(diff.mean())
    std_err = float(diff.std())
    thr = mean_err + 3.0 * std_err
    hotspots = int((diff.max(axis=2) > thr).sum())

    # Concentration: what fraction of the error energy sits in hotspots.
    energy = float(diff.sum())
    hotspot_energy = float(diff[diff.max(axis=2) > thr].sum()) if hotspots else 0.0
    concentration = hotspot_energy / energy if energy > 0 else 0.0

    # Heuristic tamper suspicion: high spread + concentrated hotspots.
    suspicion = min(1.0, (std_err / 12.0) * 0.6 + concentration * 0.6)
    flags: list[str] = []
    if suspicion > 0.55:
        flags.append("ela_hotspots")
    if mean_err < 0.35 and std_err < 0.5:
        # Extremely uniform = possibly AI-generated / heavily processed flat image
        flags.append("ela_suspiciously_uniform")

    return {
        "mean_error": round(mean_err, 4),
        "error_std": round(std_err, 4),
        "hotspot_pixels": hotspots,
        "hotspot_concentration": round(concentration, 4),
        "tamper_suspicion": round(suspicion, 3),
        "flags": flags,
    }


# ---------------------------------------------------------------------------
# Full forensic assessment
# ---------------------------------------------------------------------------

def assess_image(data: bytes, submitted_at: str | None = None) -> dict[str, Any]:
    """All forensic signals for one evidence photo."""
    img = load_image(data)
    w, h = img.size
    exif = assess_exif(img, submitted_at)
    ela = error_level_analysis(img)
    ph = perceptual_hash(img)

    # Tiny images (thumbnails) and screenshots are weak evidence.
    flags = list(exif["flags"])
    if w < 200 or h < 200:
        flags.append("low_resolution")
        exif["integrity"] = round(exif["integrity"] * 0.7, 3)
    if (w, h) in ((1080, 1920), (1080, 2400), (1170, 2532), (1284, 2778)):
        flags.append("screen_dimensions")

    return {
        "width": w,
        "height": h,
        "exif": exif,
        "ela": ela,
        "phash": ph,
        "flags": sorted(set(flags)),
    }


# ---------------------------------------------------------------------------
# Composite trust score
# ---------------------------------------------------------------------------

def _clamp01(v: float) -> float:
    return max(0.0, min(1.0, v))


def compose_trust_score(
    *,
    forensics: dict[str, Any] | None,
    cnn_category: str | None,
    cnn_severity_confidence: float | None,
    text_category: str | None,
    lat: float | None,
    lng: float | None,
    source: str = "manual",
    report_count: int = 1,
    known_phashes: list[str] | None = None,
    has_text: bool = True,
) -> dict[str, Any]:
    """Fuse forensic + contextual signals into a 0–1 trust score.

    Weights (sum to 1.0, documented so the UI can render the breakdown):
      image forensics 0.35 · image-text consistency 0.20 · geolocation 0.15
      crowd corroboration 0.15 · source reputation 0.10 · duplicate evidence 0.05

    Every component degrades gracefully — no photo, no CNN, no location:
    the missing signal's weight redistributes and the flag list says so.
    """
    flags: list[str] = []
    breakdown: dict[str, float] = {}

    # 1) Image forensics ------------------------------------------------------
    if forensics is not None:
        f_exif = forensics["exif"]["integrity"]
        f_ela = 1.0 - forensics["ela"]["tamper_suspicion"]
        forensic_signal = 0.55 * f_exif + 0.45 * f_ela
        flags.extend(forensics["flags"])
        breakdown["image_forensics"] = round(forensic_signal, 3)
        w_forensics = 0.35
    else:
        forensic_signal = 0.5  # neutral prior; weight redistributes
        breakdown["image_forensics"] = 0.5
        w_forensics = 0.0
        flags.append("no_evidence_photo")

    # 2) Image-text consistency (CNN category vs text classifier category) ----
    if cnn_category and text_category:
        consistent = cnn_category.lower() == text_category.lower()
        consistency = 1.0 if consistent else 0.15
        conf = cnn_severity_confidence if cnn_severity_confidence is not None else 0.6
        consistency = 0.3 + 0.7 * consistency * _clamp01(conf)
        if not consistent:
            flags.append(f"image_text_mismatch:{cnn_category}_vs_{text_category}")
        breakdown["image_text_consistency"] = round(consistency, 3)
        w_consistency = 0.20
    else:
        consistency = 0.5
        breakdown["image_text_consistency"] = 0.5
        w_consistency = 0.0
        if not cnn_category:
            flags.append("cnn_unavailable")

    # 3) Geolocation ----------------------------------------------------------
    geo = 0.4  # neutral when absent
    if lat is not None and lng is not None:
        # Inside plausible bounds (roughly India-wide box; demo honesty note:
        # a real deployment would use the city bounding box).
        if 6.0 <= lat <= 37.5 and 68.0 <= lng <= 97.5:
            geo = 0.9
        else:
            geo = 0.3
            flags.append("geo_out_of_region")
    else:
        flags.append("no_geolocation")
    breakdown["geolocation"] = round(geo, 3)

    # 4) Crowd corroboration --------------------------------------------------
    crowd = _clamp01(math.log1p(max(0, report_count - 1)) / math.log(12))
    breakdown["crowd_corroboration"] = round(crowd, 3)

    # 5) Source reputation ----------------------------------------------------
    rep = SOURCE_REPUTATION.get(source, 0.5)
    breakdown["source_reputation"] = round(rep, 3)

    # 6) Duplicate evidence (pHash vs previously seen photos) -----------------
    dup_signal = 1.0
    if forensics and known_phashes:
        best = max((phash_similarity(forensics["phash"], k) for k in known_phashes),
                   default=0.0)
        if best >= 0.94:
            dup_signal = 0.05
            flags.append("photo_reused:phash_match")
        elif best >= 0.86:
            dup_signal = 0.5
            flags.append("photo_similar:possible_reuse")
    breakdown["duplicate_check"] = round(dup_signal, 3)
    if not has_text:
        flags.append("no_text")

    # Weighted fusion with redistribution of unavailable-signal weight.
    weights = {
        "image_forensics": w_forensics,
        "image_text_consistency": w_consistency,
        "geolocation": 0.15,
        "crowd_corroboration": 0.15,
        "source_reputation": 0.10,
        "duplicate_check": 0.05,
    }
    active = {k: v for k, v in weights.items() if v > 0}
    total_w = sum(active.values())
    score = sum(breakdown[k] * (w / total_w) for k, w in active.items())

    # Hard penalties for hard flags.
    if any(f.startswith("photo_reused") for f in flags):
        score = min(score, 0.25)
    if any(f.startswith("edited_software") for f in flags):
        score = min(score, 0.55)
    if "ela_hotspots" in flags:
        score = min(score, 0.6)

    return {
        "score": round(_clamp01(score), 3),
        "band": trust_band(score),
        "breakdown": breakdown,
        "flags": sorted(set(flags)),
    }


def trust_band(score: float) -> str:
    if score >= 0.75:
        return "high"
    if score >= 0.5:
        return "medium"
    if score >= 0.3:
        return "low"
    return "untrusted"

File: src/lib/vision/forensics.ts
/**
 * Image forensics — pure TypeScript fallbacks mirroring ml/forensics.py.
 *
 * Three independent signals per evidence photo:
 *   1. EXIF integrity  — camera identity, capture timestamp, GPS presence,
 *                        editing-software tags, stripped-metadata tell.
 *   2. Error-Level Analysis — simulated via 8×8 block error variance: tampered
 *                        regions compress differently from their surroundings.
 *   3. DCT perceptual hash — 64-bit pHash so the same photo submitted twice is
 *                        recognized even after resize/recompression.
 *
 * These functions run in the browser (client-side preview before submit) and
 * on the server as a degraded path when the Python forensics service is
 * unreachable. The Python implementation remains the authority: richer ELA
 * (true JPEG re-encode) and full IFD parsing.
 */
import { CATEGORIES } from "../civic";

// ---------------------------------------------------------------------------
// pHash: 32×32 DCT-II, keep low 8×8, threshold at median of AC terms
// ---------------------------------------------------------------------------

const P = 32;
const LOW = 8;

/** Precomputed DCT-II orthonormal basis: C[k][n] = c(k) cos((2n+1)kπ/2N). */
function buildDctMatrix(n: number): number[][] {
  const m: number[][] = [];
  for (let k = 0; k < n; k++) {
    const row: number[] = [];
    const ck = k === 0 ? Math.sqrt(1 / n) : Math.sqrt(2 / n);
    for (let x = 0; x < n; x++) row.push(ck * Math.cos(((2 * x + 1) * k * Math.PI) / (2 * n)));
    m.push(row);
  }
  return m;
}

const DCT = buildDctMatrix(P);

/**
 * 64-bit DCT pHash as 16 hex chars. `pixels` row-major [0,1], any size.
 * Mirrors forensics.py::perceptual_hash (median over non-DC terms).
 */
export function perceptualHash(pixels: Float64Array, w: number, h: number): string {
  // Downsample to 32×32 by area-averaging.
  const small = new Float64Array(P * P);
  for (let i = 0; i < P; i++) {
    const sy0 = Math.floor((i * h) / P);
    const sy1 = Math.max(sy0 + 1, Math.floor(((i + 1) * h) / P));
    for (let j = 0; j < P; j++) {
      const sx0 = Math.floor((j * w) / P);
      const sx1 = Math.max(sx0 + 1, Math.floor(((j + 1) * w) / P));
      let sum = 0;
      let n = 0;
      for (let y = sy0; y < sy1; y++) for (let x = sx0; x < sx1; x++) { sum += pixels[y * w + x]; n++; }
      small[i * P + j] = n ? sum / n : 0;
    }
  }

  // 2D DCT via row transforms then column transforms.
  const tmp = new Float64Array(P * P);
  for (let i = 0; i < P; i++) {
    for (let k = 0; k < P; k++) {
      let sum = 0;
      for (let x = 0; x < P; x++) sum += DCT[k][x] * small[i * P + x];
      tmp[i * P + k] = sum;
    }
  }
  const freq = new Float64Array(P * P);
  for (let k = 0; k < P; k++) {
    for (let j = 0; j < P; j++) {
      let sum = 0;
      for (let y = 0; y < P; y++) sum += DCT[k][y] * tmp[y * P + j];
      freq[k * P + j] = sum;
    }
  }

  // Median of the 63 AC terms in the top-left 8×8.
  const ac: number[] = [];
  for (let k = 0; k < LOW; k++) for (let l = 0; l < LOW; l++) if (k + l > 0) ac.push(freq[k * P + l]);
  ac.sort((a, b) => a - b);
  const med = ac[ac.length >> 1];

  let value = 0;
  for (let k = 0; k < LOW; k++) {
    for (let l = 0; l < LOW; l++) {
      value = (value << 1) | (freq[k * P + l] > med ? 1 : 0);
    }
  }
  return value.toString(16).padStart(16, "0");
}

export function hammingDistance(h1: string, h2: string): number {
  let dist = 0;
  for (let i = 0; i < 16; i++) {
    let x =
      parseInt(h1[i] ?? "0", 16) ^ parseInt(h2[i] ?? "0", 16);
    while (x) {
      dist += x & 1;
      x >>= 1;
    }
  }
  return dist;
}

export function phashSimilarity(h1: string, h2: string): number {
  return 1 - hammingDistance(h1, h2) / 64;
}

// ---------------------------------------------------------------------------
// EXIF: minimal JPEG APP1/Exif parser (TIFF structure) — enough for integrity
// signals: Make, Model, Software, DateTime(Original), GPS IFD presence.
// ---------------------------------------------------------------------------

export type ExifSummary = {
  make: string | null;
  model: string | null;
  software: string | null;
  capturedAt: string | null;
  hasGps: boolean;
  tagCount: number;
};

const EDITOR_SOFTWARE = [
  "photoshop", "adobe", "lightroom", "gimp", "snapseed", "canva",
  "picsart", "facetune", "pixelmator", "affinity", "illustrator",
  "after effects", "figma", "paint.net", "paint 3d", "mspaint",
  "krita", "darktable", "polarr", "vsco", "meitu", "b612",
];

/** Returns null for non-JPEG inputs (PNG/WebP typically carry no EXIF in browsers). */
export function parseExif(buf: ArrayBuffer): ExifSummary | null {
  const v = new DataView(buf);
  if (buf.byteLength < 4 || v.getUint16(0) !== 0xffd8) return null; // not JPEG

  let offset = 2;
  while (offset + 4 < buf.byteLength) {
    if (v.getUint8(offset) !== 0xff) break;
    const marker = v.getUint8(offset + 1);
    const size = v.getUint16(offset + 2);
    if (marker === 0xe1 && offset + 4 + 6 <= buf.byteLength) {
      const sig = String.fromCharCode(...new Uint8Array(buf, offset + 4, 6));
      if (sig === "Exif\0\0") {
        return parseTiff(buf, offset + 10);
      }
    }
    offset += 2 + size;
  }
  return { make: null, model: null, software: null, capturedAt: null, hasGps: false, tagCount: 0 };
}

function parseTiff(buf: ArrayBuffer, base: number): ExifSummary | null {
  try {
    const v = new DataView(buf);
    const bom = v.getUint16(base);
    const little = bom === 0x4949;
    if (!little && bom !== 0x4d4d) return null;
    const u16 = (o: number) => v.getUint16(base + o, little);
    const u32 = (o: number) => v.getUint32(base + o, little);

    const ifd0Off = u32(4);
    const out: ExifSummary = { make: null, model: null, software: null, capturedAt: null, hasGps: false, tagCount: 0 };
    let exifIfdOff = 0;
    let gpsIfdOff = 0;

    const readIfd = (off: number): number => {
      const count = u16(off);
      out.tagCount += count;
      for (let i = 0; i < count; i++) {
        const e = off + 2 + i * 12;
        const tag = u16(e);
        const type = u16(e + 2);
        const num = u32(e + 4);
        const valOff = num <= 4 ? e + 8 : base + u32(e + 8);
        const readStr = () => {
          const bytes = new Uint8Array(buf, valOff, Math.min(num, 128));
          return new TextDecoder("ascii").decode(bytes).replace(/\0.*$/, "").trim();
        };
        if (tag === 0x010f) out.make = readStr();
        else if (tag === 0x0110) out.model = readStr();
        else if (tag === 0x0131) out.software = readStr();
        else if (tag === 0x0132) out.capturedAt = readStr();
        else if (tag === 0x8769) exifIfdOff = u32(e + 8);
        else if (tag === 0x8825) gpsIfdOff = u32(e + 8);
      }
      return u32(off + 2 + count * 12);
    };

    let next = readIfd(ifd0Off);
    if (exifIfdOff) {
      const e = exifIfdOff;
      const cnt = u16(e);
      out.tagCount += cnt;
      for (let i = 0; i < cnt; i++) {
        const ent = e + 2 + i * 12;
        const tag = u16(ent);
        const num = u32(ent + 4);
        const valOff = num <= 4 ? ent + 8 : base + u32(ent + 8);
        if (tag === 0x9003 || tag === 0x9004) {
          const bytes = new Uint8Array(buf, valOff, Math.min(num, 32));
          out.capturedAt = new TextDecoder("ascii").decode(bytes).replace(/\0.*$/, "").trim();
        }
      }
      void next;
    }
    if (gpsIfdOff) {
      out.hasGps = true;
      const cnt = u16(gpsIfdOff);
      out.tagCount += cnt;
    }
    return out;
  } catch {
    return null;
  }
}

export type ExifAssessment = {
  integrity: number;
  flags: string[];
  editedBy: string | null;
  hasCamera: boolean;
  hasTimestamp: boolean;
  hasGps: boolean;
};

export function assessExifJs(summary: ExifSummary | null): ExifAssessment {
  const flags: string[] = [];
  if (!summary) {
    return { integrity: 0.2, flags: ["no_exif_or_non_jpeg"], editedBy: null, hasCamera: false, hasTimestamp: false, hasGps: false };
  }
  const blob = `${summary.software ?? ""} ${summary.make ?? ""} ${summary.model ?? ""}`.toLowerCase();
  let editedBy: string | null = null;
  for (const ed of EDITOR_SOFTWARE) {
    if (blob.includes(ed)) { editedBy = ed; break; }
  }
  const hasCamera = Boolean(summary.make || summary.model);
  const hasTimestamp = Boolean(summary.capturedAt);
  const hasGps = summary.hasGps;

  if (editedBy) flags.push(`edited_software:${editedBy}`);
  if (summary.tagCount === 0) flags.push("metadata_stripped");
  else if (!hasCamera) flags.push("no_camera_identity");
  if (!hasTimestamp) flags.push("no_capture_timestamp");
  if (!hasGps) flags.push("no_gps_exif");

  let integrity = 0;
  if (hasCamera) integrity += 0.3;
  if (hasTimestamp) integrity += 0.3;
  if (hasGps) integrity += 0.2;
  if (!editedBy) integrity += 0.2;
  return { integrity: Math.min(1, integrity), flags, editedBy, hasCamera, hasTimestamp, hasGps };
}

// ---------------------------------------------------------------------------
// Block error analysis (ELA surrogate): JPEG re-encode isn't available
// server-side in TS, so we measure high-frequency energy concentration.
// Spliced regions have atypically sharp block-boundary discontinuities.
// ---------------------------------------------------------------------------

export type ElaAssessment = {
  meanError: number;
  errorStd: number;
  hotspotConcentration: number;
  tamperSuspicion: number;
  flags: string[];
};

export function blockErrorAnalysis(pixels: Float64Array, w: number, h: number): ElaAssessment {
  // Laplacian-style high-pass response as a proxy for recompression error.
  const resp: number[] = [];
  for (let y = 1; y < h - 1; y++) {
    for (let x = 1; x < w - 1; x++) {
      const c = pixels[y * w + x];
      const lap = 4 * c - pixels[y * w + x - 1] - pixels[y * w + x + 1] - pixels[(y - 1) * w + x] - pixels[(y + 1) * w + x];
      resp.push(Math.abs(lap));
    }
  }
  const mean = resp.reduce((a, b) => a + b, 0) / Math.max(1, resp.length);
  const variance = resp.reduce((a, b) => a + (b - mean) ** 2, 0) / Math.max(1, resp.length);
  const std = Math.sqrt(variance);
  const thr = mean + 3 * std;
  const hot = resp.filter((v) => v > thr).length;
  const concentration = resp.length ? hot / resp.length : 0;
  const suspicion = Math.min(1, (std / 12) * 0.6 + concentration * 6);
  const flags: string[] = [];
  if (suspicion > 0.55) flags.push("ela_hotspots");
  if (mean < 0.01 && std < 0.02) flags.push("ela_suspiciously_uniform");
  return {
    meanError: round4(mean),
    errorStd: round4(std),
    hotspotConcentration: round4(concentration),
    tamperSuspicion: round4(suspicion),
    flags,
  };
}

function round4(v: number): number {
  return Math.round(v * 10000) / 10000;
}

// ---------------------------------------------------------------------------
// Shared trust composition — mirrors forensics.py::compose_trust_score
// ---------------------------------------------------------------------------

export type TrustInput = {
  forensics: { exifIntegrity: number; elaSuspicion: number; flags: string[] } | null;
  cnnCategory: string | null;
  cnnConfidence: number | null;
  textCategory: string | null;
  lat: number | null;
  lng: number | null;
  source: string;
  reportCount: number;
  knownPhashes: string[];
  hasText: boolean;
};

export const SOURCE_REPUTATION: Record<string, number> = {
  whatsapp: 0.75, email: 0.7, manual: 0.7, x: 0.55, tweet: 0.55,
  news: 0.8, social: 0.5, bulk: 0.4,
};

const clamp01 = (v: number) => Math.max(0, Math.min(1, v));

export function trustBand(score: number): "high" | "medium" | "low" | "untrusted" {
  if (score >= 0.75) return "high";
  if (score >= 0.5) return "medium";
  if (score >= 0.3) return "low";
  return "untrusted";
}

export function composeTrustJs(input: TrustInput): {
  score: number;
  band: ReturnType<typeof trustBand>;
  breakdown: Record<string, number>;
  flags: string[];
} {
  const flags: string[] = [];
  const breakdown: Record<string, number> = {};
  const weights: Record<string, number> = {
    image_forensics: 0,
    image_text_consistency: 0,
    geolocation: 0.15,
    crowd_corroboration: 0.15,
    source_reputation: 0.1,
    duplicate_check: 0.05,
  };

  if (input.forensics) {
    const forensicSignal = 0.55 * input.forensics.exifIntegrity + 0.45 * (1 - input.forensics.elaSuspicion);
    breakdown.image_forensics = Math.round(forensicSignal * 1000) / 1000;
    weights.image_forensics = 0.35;
    flags.push(...input.forensics.flags);
  } else {
    breakdown.image_forensics = 0.5;
    flags.push("no_evidence_photo");
  }

  if (input.cnnCategory && input.textCategory) {
    const consistent = input.cnnCategory.toLowerCase() === input.textCategory.toLowerCase();
    let consistency = 1.0;
    if (!consistent) {
      consistency = 0.15;
      flags.push(`image_text_mismatch:${input.cnnCategory}_vs_${input.textCategory}`);
    }
    const conf = clamp01(input.cnnConfidence ?? 0.6);
    breakdown.image_text_consistency = Math.round((0.3 + 0.7 * consistency * conf) * 1000) / 1000;
    weights.image_text_consistency = 0.2;
  } else {
    breakdown.image_text_consistency = 0.5;
    if (!input.cnnCategory) flags.push("cnn_unavailable");
  }

  let geo = 0.4;
  if (input.lat != null && input.lng != null) {
    if (input.lat >= 6.0 && input.lat <= 37.5 && input.lng >= 68.0 && input.lng <= 97.5) geo = 0.9;
    else { geo = 0.3; flags.push("geo_out_of_region"); }
  } else flags.push("no_geolocation");
  breakdown.geolocation = geo;

  breakdown.crowd_corroboration =
    Math.round(clamp01(Math.log1p(Math.max(0, input.reportCount - 1)) / Math.log(12)) * 1000) / 1000;

  breakdown.source_reputation = SOURCE_REPUTATION[input.source] ?? 0.5;

  let dupSignal = 1.0;
  if (input.forensics && input.knownPhashes.length) {
    // The caller supplies the current photo's pHash via forensics.flags side-
    // channel is not possible — instead knownPhashes are compared by the
    // caller, who passes the resulting duplicate penalty through flags.
    void input.knownPhashes;
  }
  breakdown.duplicate_check = dupSignal;
  if (!input.hasText) flags.push("no_text");

  const active = Object.entries(weights).filter(([, w]) => w > 0);
  const totalW = active.reduce((a, [, w]) => a + w, 0);
  let score = active.reduce((a, [k, w]) => a + breakdown[k] * (w / totalW), 0);

  if (flags.some((f) => f.startsWith("photo_reused"))) score = Math.min(score, 0.25);
  if (flags.some((f) => f.startsWith("edited_software"))) score = Math.min(score, 0.55);
  if (flags.includes("ela_hotspots")) score = Math.min(score, 0.6);

  return {
    score: Math.round(clamp01(score) * 1000) / 1000,
    band: trustBand(score),
    breakdown,
    flags: [...new Set(flags)].sort(),
  };
}

// Re-export so callers can reference categories without importing civic twice.
export { CATEGORIES as CNN_CATEGORY_LIST };
